icu4c/source/i18n/calendar.cpp

```diff
@@ -449,7 +449,7 @@ class BasicCalendarFactory : public LocaleKeyFactory {
             fprintf(stderr, "::create - not a LocaleKey!\n");
         }
 #endif
-        const LocaleKey& lkey = (LocaleKey&)key;
+        const LocaleKey& lkey = dynamic_cast<const LocaleKey&>(key);
```

Comment:
Again we should not need to change any C-style cast to a dynamic_cast.
I won't comment on more of these below :-)

Reply:
If I change this line to static_cast<> I got the following runtime error during CFI
```
         TestGenericAPI {
calendar.cpp:452:33: runtime error: control flow integrity check for type 'icu::LocaleKey' failed during base-to-derived cast (vtable address 0x7fc2e45fa730)
0x7fc2e45fa730: note: invalid vtable
 c2 7f 00 00  a0 7e 4c e4 c2 7f 00 00  80 7b 4c e4 c2 7f 00 00  a0 7b 4c e4 c2 7f 00 00  f0 59 4c e4
              ^ 
calendar.cpp:452:33: note: check failed in ../../lib/libicui18n.so.73, vtable located in ../../lib/libicuuc.so.73
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior calendar.cpp:452:33 in 

```